sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_read.h>
+
+#include <az_digit.h>
+
+#include <math.h>
+
+inline bool az_json_is_white_space(char const c) {
+  switch (c) {
+    case ' ': case '\t': case '\n': case '\r':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_esc(char const c) {
+  switch (c) {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_stack_is_empty(az_json_state const *const p) {
+  return p->stack == 1;
+}
+
+inline az_json_stack_item az_json_stack_last(az_json_state const *const p) {
+  return p->stack & 1;
+}
+
+az_json_state az_json_state_create(az_const_str const buffer) {
+  return (az_json_state){
+    .buffer = buffer,
+    .i = 0,
+    .stack = 1,
+  };
+}
+
+az_result az_json_get_char(az_const_str const buffer, size_t const i, char *const out_char) {
+  if (i == buffer.size) {
+    return AZ_JSON_ERROR_UNEXPECTED_END;
+  }
+  *out_char = az_const_str_item(buffer, i);
+  return AZ_OK;
+}
+
+az_result az_json_expect_char(az_const_str const buffer, size_t *const p, char const expected) {
+  char c;
+  AZ_RETURN_ON_ERROR(az_json_get_char(buffer, *p, &c));
+  if (c != expected) {
+    return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+  }
+  ++*p;
+  return AZ_OK;
+}
+
+az_result az_json_read_white_space(az_const_str const buffer, size_t *const p) {
+  for (; *p != buffer.size; ++*p) {
```

Comment:
I'm not sure I understand what p is supposed to be pointing to here? and *p != buffer.size looks suspicious at first glance, I don't think it's a bug, but it's confusing and could use a comment.

I suspect that just having the parameter be `size_t p` then for(; p != buffer.size; ++p)` would be better.

Reply:
it moves the position `*p` to the first character after whitespace. Currently, a pair of `az_const_str buffer` and `size_t *const p_position` is a kind of read stream.

Reply:
hmm still don't love it, but OK.

Reply:
We have several options.
1. Use a local variable and then assign it back. (It's quite error prone, I did forget several times to assigned it back).
2. Use a function result, but it requires complex structures and change error conventions.
3. Use in/out parameter (as it is now)
4. Use new structure, something like 
```c
typedef struct {
  az_const_str const buffer;
  size_t position;
} az_const_str_reader;
```